src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KeyAgreement.java

```diff
@@ -268,20 +268,18 @@ protected SecretKey engineGenerateSecret(String algorithm)
             throw new NoSuchAlgorithmException("Algorithm must not be null");
         }
 
-        if (algorithm.equals("TlsPremasterSecret")) {
+        if (KeyUtil.isSupportedKeyAgreementOutputAlgorithm(algorithm)) {
             // For now, only perform native derivation for TlsPremasterSecret
-            // as that is required for FIPS compliance.
+            // as that is required for FIPS compliance and Generic for KDF.
```

Comment:
Suggest breaking into two sentences: "For now, only perform native derivation for TlsPremasterSecret and Generic algorithms. TlsPremasterSecret is required for FIPS compliance and Generic is required for input to KDF."